### PR TITLE
Add fit parameters passing 

### PR DESCRIPTION
**Pull request by sami-ka**
# Description

In the MAPIE code, the fit function does not allow to pass parameters usable at fit time for base estimator. This is the case for early stopping in LightGBM for instance.

Fixes partially #212 (fit_params only)

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] Add test with passing `monitor `parameter in `fit` function of `GradientBoostingRegressor`
- [x] Check that the other tests still work

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

File: mapie/regression/quantile_regression.py [hunk 1]
```diff
@@ -609,8 +613,12 @@ def fit(
...
                 self.estimators_.append(fit_estimator(
-                    cloned_estimator_, X_train, y_train, sample_weight_train
-                ))
+                    cloned_estimator_,
+                    X_train,
+                    y_train,
+                    sample_weight_train,
+                    **fit_params,
+                    ))
                 y_calib_preds[i] = self.estimators_[-1].predict(X_calib)
```
Comment (thibaultcordier) [inline mapie/regression/quantile_regression.py:622]: Just a writing style suggestion:
```suggestion
                ))
```

File: mapie/regression/regression.py [hunk 2]
```diff
@@ -507,7 +511,9 @@ def fit(
             self.verbose
         )
         # Fit the prediction function
-        self.estimator_ = self.estimator_.fit(X, y, sample_weight)
+        self.estimator_ = self.estimator_.fit(
+            X, y, sample_weight, **fit_params
+            )
```
Comment (thibaultcordier) [inline mapie/regression/regression.py:516]: Just a writing style suggestion:
```suggestion
        )
```

File: mapie/classification.py [hunk 3]
```diff
@@ -633,11 +637,13 @@ def _fit_and_predict_oof_model(
         y_val = _safe_indexing(y, val_index)
 
         if sample_weight is None:
-            estimator = fit_estimator(estimator, X_train, y_train)
+            estimator = fit_estimator(
+                estimator, X_train, y_train, **fit_params
+                )
```
Comment (thibaultcordier) [inline mapie/classification.py:642]: Just a writing style suggestion:
```suggestion
            )
```

**Review by LacombeLouis — CHANGES_REQUESTED** [review]
Hi @sami-ka,
This is great! Thank you for contributing to MAPIE.
I see that you've added for all the instances for `**fit_params`.
One small issue is that you've currently only implemented one test. I suggest adding tests for classification and regression.
Thank you!

File: mapie/tests/test_quantile_regression.py [hunk 4]
```diff
@@ -762,3 +771,28 @@ def test_consistent_class() -> None:
...
+        X_train,
+        y_train,
+        X_calib=X_calib,
+        y_calib=y_calib,
+        sample_weight=None,
+        monitor=early_stopping_monitor
+        )
+
+    for estimator in mapie.estimators_:
+        assert estimator.estimators_.shape[0] == 3
```
Comment (thibaultcordier) [inline mapie/tests/test_quantile_regression.py:798]: I'm not sure I understand what's being tested here. By construction, we should have `estimator.estimators_.shape[0] == GradientBoostingRegressor.n_estimators` (`n_estimators int, default=100`) which corresponds to the number of estimators.`monitor=early_stopping_monitor` forces the maximum number of estimators, have I understood correctly?

File: requirements.dev.txt [hunk 5]
```diff
@@ -2,7 +2,7 @@ bump2version==1.0.1
 flake8==4.0.1
 ipykernel==6.9.0
 jupyter==1.0.0
-mypy==0.941
+mypy==0.981
```
Comment (thibaultcordier) [inline requirements.dev.txt:5]: This line should not be changed. What justifies this change? Did you encounter any problems when you ran your tests?

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
Thank you very much for this proposal, which will extend the use of MAPIE to more estimators! Here is my feedback to complete this PR:
- add unit tests for calibration, classification and regression (as you did for quantile regression). No need for time series.
- some writing style suggestion.

File: mapie/tests/test_quantile_regression.py (hunk 4 above)
Comment (sami-ka) [reply mapie/tests/test_quantile_regression.py:798]: What you see here is the usage of early stopping in boosting algorithms of scikit learn (https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingRegressor.html#sklearn.ensemble.GradientBoostingRegressor.fit).
What is usually done in boosting is to look at the variation of a metric on a validation set (and not the training set), and to stop when this metric is not improving for a defined number of iterations, and thus to stop before reaching the maximum number of estimators (which is indeed 100 by default). That limits the risk of overfitting. You also have that mechanism in neural network for instance.
For test purposes, I looked at what scikit-learn does currently, which is to use a dummy early stopping procedure that stops at a defined number of iteration (https://github.com/scikit-learn/scikit-learn/blob/4ce8e19859cb8b2f2bef197ed5b28beea44ee4b4/sklearn/ensemble/tests/test_gradient_boosting.py#L981). Here I chose 3 in the `early_stopping_monitor` function. That's why I test that the number of estimators used in boosting is indeed 3 instead of 100, which means the additional fit parameters have been taken into account.

Comment (thibaultcordier) [reply mapie/tests/test_quantile_regression.py:798]: Very clear, thank you for these details which you can add in the test docstring (2 sentences) so that no one forgets.

File: mapie/tests/test_calibration.py [hunk 6]
```diff
@@ -106,6 +106,18 @@
     X_, y_, test_size=0.33, random_state=random_state
 )
 
+gb = GradientBoostingClassifier(
+            random_state=random_state
+            )
```
Comment (thibaultcordier) [inline mapie/tests/test_calibration.py:111]:
```suggestion
gb = GradientBoostingClassifier(random_state=random_state)
```

File: mapie/tests/test_calibration.py [hunk 7]
```diff
@@ -461,3 +473,20 @@ def test_pipeline_compatibility() -> None:
...
+    Checks that underlying GradientBoosting estimators have used 3 iterations
+    only during boosting, instead of default value for n_estimators (=100).
+    """
+    mapie = MapieCalibrator(estimator=gb)
+
+    mapie.fit(
+        X,
+        y,
+        monitor=early_stopping_monitor
+        )
```
Comment (thibaultcordier) [inline mapie/tests/test_calibration.py:490]:
```suggestion
    mapie.fit(X, y, monitor=early_stopping_monitor)
```

File: mapie/tests/test_classification.py [hunk 8]
```diff
@@ -821,6 +822,18 @@
     random_state=random_state,
 )
 
+gb = GradientBoostingClassifier(
+            random_state=random_state
+            )
```
Comment (thibaultcordier) [inline mapie/tests/test_classification.py:827]:
```suggestion
gb = GradientBoostingClassifier(random_state=random_state)
```

File: mapie/tests/test_classification.py [hunk 9]
```diff
@@ -1957,3 +1970,26 @@ def test_deprecated_method_warning(method: str) -> None:
...
+def test_fit_parameters_passing() -> None:
+    """
+    Test passing fit parameters, here early stopping at iteration 3.
+    Checks that underlying GradientBoosting estimators have used 3 iterations
+    only during boosting, instead of default value for n_estimators (=100).
+    """
+    mapie = MapieClassifier(
+        estimator=gb,
+        method="aps"
+        )
```
Comment (thibaultcordier) [inline mapie/tests/test_classification.py:1984]:
```suggestion
    mapie = MapieClassifier(
        estimator=gb, method="aps", random_state=random_state
    )
```

File: mapie/tests/test_calibration.py [hunk 10]
```diff
@@ -461,3 +473,20 @@ def test_pipeline_compatibility() -> None:
...
     mapie.predict(X)
+
+
+def test_fit_parameters_passing() -> None:
+    """
+    Test passing fit parameters, here early stopping at iteration 3.
+    Checks that underlying GradientBoosting estimators have used 3 iterations
+    only during boosting, instead of default value for n_estimators (=100).
+    """
+    mapie = MapieCalibrator(estimator=gb)
```
Comment (thibaultcordier) [inline mapie/tests/test_calibration.py:484]:
```suggestion
    mapie = MapieCalibrator(estimator=gb, random_state=random_state)
```

File: mapie/tests/test_classification.py [hunk 11]
```diff
@@ -1957,3 +1970,26 @@ def test_deprecated_method_warning(method: str) -> None:
...
+
+    mapie.fit(
+        X,
+        y,
+        monitor=early_stopping_monitor
+        )
```
Comment (thibaultcordier) [inline mapie/tests/test_classification.py:1990]:
```suggestion
    mapie.fit(X, y, monitor=early_stopping_monitor)
```

File: mapie/tests/test_quantile_regression.py [hunk 12]
```diff
@@ -762,3 +771,30 @@ def test_consistent_class() -> None:
...
+) -> None:
+    """
+    Test passing fit parameters, here early stopping at iteration 3.
+    Checks that underlying GradientBoosting estimators have used 3 iterations
+    only during boosting, instead of default value for n_estimators (=100).
+    """
+    mapie = MapieQuantileRegressor(
+        estimator=gb,
+        **STRATEGIES[strategy]
+        )
```
Comment (thibaultcordier) [inline mapie/tests/test_quantile_regression.py:788]:
```suggestion
    mapie = MapieQuantileRegressor(estimator=gb, **STRATEGIES[strategy])
```

File: mapie/tests/test_quantile_regression.py [hunk 13]
```diff
@@ -762,3 +771,30 @@ def test_consistent_class() -> None:
...
+
+    mapie.fit(
+        X_train,
+        y_train,
+        X_calib=X_calib,
+        y_calib=y_calib,
+        sample_weight=None,
+        monitor=early_stopping_monitor
+        )
```
Comment (thibaultcordier) [inline mapie/tests/test_quantile_regression.py:797]:
```suggestion
    )
```

File: mapie/tests/test_regression.py [hunk 14]
```diff
@@ -39,6 +40,19 @@
 
 random_state = 1
 
+gb = GradientBoostingRegressor(
+            random_state=random_state
+            )
```
Comment (thibaultcordier) [inline mapie/tests/test_regression.py:45]:
```suggestion
gb = GradientBoostingRegressor(random_state=random_state)
```

File: mapie/tests/test_regression.py [hunk 15]
```diff
@@ -646,3 +660,25 @@ def test_beta_optimize_user_warning() -> None:
...
+
+def test_fit_parameters_passing() -> None:
+    """
+    Test passing fit parameters, here early stopping at iteration 3.
+    Checks that underlying GradientBoosting estimators have used 3 iterations
+    only during boosting, instead of default value for n_estimators (=100).
+    """
+    mapie = MapieRegressor(
+                estimator=gb,
+            )
```
Comment (thibaultcordier) [inline mapie/tests/test_regression.py:673]:
```suggestion
    mapie = MapieRegressor(estimator=gb, random_state=random_state)
```

File: mapie/tests/test_regression.py [hunk 16]
```diff
@@ -646,3 +660,25 @@ def test_beta_optimize_user_warning() -> None:
...
+
+    mapie.fit(
+        X,
+        y,
+        monitor=early_stopping_monitor
+        )
```
Comment (thibaultcordier) [inline mapie/tests/test_regression.py:679]:
```suggestion
    mapie.fit(X, y, monitor=early_stopping_monitor)
```

File: mapie/tests/test_calibration.py (hunk 6 above)
Comment (thibaultcordier) [reply mapie/tests/test_calibration.py:111]: I also suggest moving it directly into the test (before instantiating the `mapie` estimator) to better understand where `gb` is being used.

File: mapie/tests/test_calibration.py [hunk 17]
```diff
@@ -106,6 +106,18 @@
...
+            random_state=random_state
+            )
+
+
+def early_stopping_monitor(i, est, locals):
+    """Returns True on the 3rd iteration."""
+    if i == 2:
+        return True
+    else:
+        return False
```
Comment (thibaultcordier) [inline mapie/tests/test_calibration.py:119]: I also suggest moving it directly into the test to better understand where the function is being used.

File: mapie/tests/test_classification.py [hunk 18]
```diff
@@ -821,6 +822,18 @@
...
+            random_state=random_state
+            )
+
+
+def early_stopping_monitor(i, est, locals):
+    """Returns True on the 3rd iteration."""
+    if i == 2:
+        return True
+    else:
+        return False
```
Comment (thibaultcordier) [inline mapie/tests/test_classification.py:835]: I also suggest moving it directly into the test to better understand where the function is being used.

File: mapie/tests/test_quantile_regression.py [hunk 19]
```diff
@@ -46,6 +46,15 @@
             random_state=random_state
             )
 
+
+def early_stopping_monitor(i, est, locals):
+    """Returns True on the 3rd iteration."""
+    if i == 2:
+        return True
+    else:
+        return False
```
Comment (thibaultcordier) [inline mapie/tests/test_quantile_regression.py:55]: I also suggest moving it directly into the test to better understand where the function is being used.

File: mapie/tests/test_regression.py [hunk 20]
```diff
@@ -39,6 +40,19 @@
...
+            random_state=random_state
+            )
+
+
+def early_stopping_monitor(i, est, locals):
+    """Returns True on the 3rd iteration."""
+    if i == 2:
+        return True
+    else:
+        return False
```
Comment (thibaultcordier) [inline mapie/tests/test_regression.py:53]: I also suggest moving it directly into the test to better understand where the function is being used.

File: mapie/tests/test_classification.py (hunk 8 above)
Comment (thibaultcordier) [reply mapie/tests/test_classification.py:827]: I also suggest moving it directly into the test (before instantiating the `mapie` estimator) to better understand where `gb` is being used.

File: mapie/tests/test_regression.py (hunk 14 above)
Comment (thibaultcordier) [reply mapie/tests/test_regression.py:45]: I also suggest moving it directly into the test (before instantiating the `mapie` estimator) to better understand where `gb` is being used.

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
All right, then! We're about to wrap up this PR. I'd like to suggest a few changes to help us maintain your code in the future.

I see that we've instantiated the `gb` object at the start of the `test_quantile_regression.py` file, but I think it would be better to do this directly in your tests in the other files (as it won't be used elsewhere except for `test_quantile_regression.py`).

**Comment by sami-ka** [comment]
> All right, then! We're about to wrap up this PR. I'd like to suggest a few changes to help us maintain your code in the future.
> 
> I see that we've instantiated the `gb` object at the start of the `test_quantile_regression.py` file, but I think it would be better to do this directly in your tests in the other files (as it won't be used elsewhere except for `test_quantile_regression.py`).

You are right, it would definitely be more relevant inside the tests. Thanks for the suggestion!

File: mapie/tests/test_calibration.py (hunk 10 above)
Comment (sami-ka) [reply mapie/tests/test_calibration.py:474]: @thibaultcordier it seems that there is no `random_state` argument in the init function of `MapieCalibrator` so I did not make this change.

**Review by vincentblot28 — APPROVED** [review]

**Comment by sami-ka** [comment]
@thibaultcordier Is there something else to do? Could this PR be merged?

**Comment by thibaultcordier** [comment]
@sami-ka I'll let @LacombeLouis confirm that he has no more change requests. He will then merge your PR.

**Review by LacombeLouis — APPROVED** [review]
